src/backend/distributed/master/master_repair_shards.c

```diff
@@ -488,3 +511,44 @@ WorkerApplyShardDDLCommandList(List *ddlCommandList, int64 shardId)
 
 	return applyDdlCommandList;
 }
+
+
+/*
+ * LockReferencedTablesIfExist takes exclusive lock on the reference tables
+ * which has a foreign key from the given relation. This function is used
+ * for shard_rebalancer and tenant isolation to avoid modifications on the
+ * referenced end of a foreign key while we run shard_rebalancer or tenant
+ * isolation on the referencing end of it.
+ * If we do not block DMLs on the referenced table, we cannot avoid the
+ * inconsistency between the two copies of the data.
+ */
+static void
+LockReferencedTablesIfExist(Oid relationId)
+{
+	List *referencedTableList = NIL;
+	ListCell *referencedTableCell = NULL;
+
+	referencedTableList = GetReferencedReferenceTableList(relationId);
+	referencedTableList = SortList(referencedTableList, CompareOids);
+
+	foreach(referencedTableCell, referencedTableList)
+	{
+		Oid referencedTableId = lfirst_oid(referencedTableCell);
+		List *shardList = NIL;
+		uint64 *shardIdPointer = NULL;
+		Oid referencedShardId = InvalidOid;
+
+		/* check that user has owner rights in referenced table */
+		EnsureTableOwner(referencedTableId);
```

Comment:
Future note: We should skip `non-reference` tables since the `foreign_key_relations` does not return only reference tables. And, we should add this information to the function comment.